### ISSUE TITLE
Investigate how plot functions might be tested

### ISSUE THREAD

**Issue by jacg** [next-exp/IC#106]
Our code base contains functions whose purpose is to create plots. These are currently untested. Options include

1. not test them at all
1. ad-hoc test them, perhaps in notebooks
1. try to write proper automated tests

@jacg is to investigate the feasibility of option 3

@jjgomezcadenas please make a list of all such functions. (I would do it myself, but given that matplotlib is sometimes used in our code to perform calculations rather than to make plots, I'm likely to get it wrong :-)

**Comment by jjgomezcadenas** [comment]
A simple possibility would be to add, for each plotting function (or group of functions) a notebook where the plot is illustrated. This is fast (the developer always tries the plotting function in the NB) and it has documentation value. 

And YES, got message, decouple plotting from calculations! (old hand-ROOTED habits).

**Comment by jacg** [comment]
Nope, that would be pretty ad-hoc. That's not what I'm after:

- Developers will not have the self-discipline or the time to do this regularly.
- Good luck persuading Travis to do it.

No, I was thinking more about mocking. Don't worry about it, it's something I have to look into when I have some time. (Though I'd be happy to talk to you about it in person. And it would probably help me to discuss these functions with you.)

**Comment by jjgomezcadenas** [comment]
As a matter of fact, I think developers will do whatever is agreed and reinforced. By the same token that they are expected to submit a test when they PR for a new function, they can be asked to submit a NB when they PR a new plotting function. And it turns out that in most cases they will have done it anyway, since they will necessarily try their plotting function in a NB (essentially 100 % of the developers will write and try the function in the NB first).

In that sense, it is no burden at all, and the function is automatically documented. Travis does not care, and does not need to, from his point of view this is just a piece of source code.

The contrary probably is also true. If developer are asked to so something more complicated than submitting a NB per plot function or group, they will tend to keep their plotting functions in their NB and skip PRs. 

Plotting functions are just utilities, in a sense their best test is that they do what they promise (e.g, plot object). Sure enough we can reinforce a separation between plotting and calculating something (as you spotted in one of my commits, later corrected), but I think is not worth to spend too much time there. We have our platter full of other stuff.

**Comment by jacg** [comment]
> As a matter of fact, I think developers will do whatever is agreed
> and reinforced.

This is most certainly **not** true.

> By the same token that they are expected to submit a test when they
> PR for a new function,

I'm glad you brought that up ... because you yourself have not
submitted tests along with new functionality on occasion. Thus we have
experimental evidence directly falsifying your hypothesis above.

Furthermore, you yourself have failed to run the test suite before
submitting a PR, once again contradicting your claim above.

Pressing that `run-tests` button is a few orders of magnitude simpler
and cheaper than actually writing new tests, and yet, somehow, a
person who agrees with and promotes this policy has repeatedly failed
to follow it. So, you can be sure that there will be plenty of
occasions on which **developers will not do what is agreed**.

> they can be asked to submit a NB when they PR a new plotting
> function.

In similar vein they can be asked to submit a test which compares the
currently generated plot with the original one. The benefit of the
latter is that it will be automatically checked by the test framework,
while your suggestion unecessarily burdens the developer with the need
to perform checks manually by eye ... which will lead to the checks
*not* being done, which effectively leaves the functions untested.

Initial investigation suggests that this is a solved problem (the
functionality comes both with matplotlib and in a pytest plugin). So
we really don't need to be having this discussion. We really don't
need to discuss how to re-invent the wheel badly. We really don't need
to champion NIH syndrome. We really don't need to champion ad-hoc testing (which is, effectively, not testing).

Once I get some spare time, I will produce a recipe for how this
should be done, and that's it.

> And it turns out that in most cases they will have done it anyway,
> since they will necessarily try their plotting function in a NB
> (essentially 100 % of the developers will write and try the function
> in the NB first). In that sense, it is no burden at all, and the
> function is automatically documented.

Be we need it to be automatically **tested**.

> Travis does not care, and does not need to, from his point of view
> this is just a piece of source code.

Travis absolutely *does* care, because it is Travis' job to test our
functionality, and Travis does not have the human intelligence
necessary to carry out this task. Travis can only build our code and
run our *automated* tests. That's why this needs to go into an
*automated* test.

If you insist on trying to persuade me that chunks of your code don't
need automatic testing, I'll re-instate coverage: you have been
warned!

> The contrary probably is also true. If developer are asked to so
> something more complicated than submitting a NB per plot function or
> group, they will tend to keep their plotting functions in their NB
> and skip PRs.

You have it backwards. The important point is: if a develpoer is asked
to do anything more complicated than presing the 'run all our
automated tests' button (for example, compare a bunch of graphs by
eye, in addition to pressing that button, like you are suggesting)
then the developers will skip that part, and the utilities will be
effectively untested.

> Plotting functions are just utilities, in a sense their best test is
> that they do what they promise (e.g, plot object).

But not just any plot object, but one that actually has some specific
desirable features. Testing this can, and should, be automated.

> Sure enough we can reinforce a separation between plotting and
> calculating something (as you spotted in one of my commits, later
> corrected),

In the first instance, that has nothing to do with testing. Separation
of concerns is simply good design. One of the (many) benefits of
separation of concerns is that testing becomes easier.

> but I think is not worth to spend too much time there. We have our
> platter full of other stuff.

Again, you have it the wrong way around: we *cannot afford* to do it
your way.

You claim to have understood the need for testing, but this whole
discussion shows that you still have a long way to go.

Hopefully I'll have the recipe ready before the weekend.

**Comment by jjgomezcadenas** [comment]
> You claim to have understood the need for testing, but this whole
> discussion shows that you still have a long way to go.
> 
> 
This whole discussion shows that I like to discuss. I am sure I have a long road ahead toward perfection (who doesn’t except our Jedi?). I also think that testing plots is less of a priority for us than may other things, and I know for a fact, we are late and behind schedule in the targets I had in mind for development. Perhaps I was simply too optimistic. But I still dare to state my opinion.
> Hopefully I'll have the recipe ready before the weekend.
> 
> 
Looking forward to it. Lately it appears my progress is slowing to a halt. I can’t even manage to get a simple PR approved.

**Comment by jacg** [comment]
> This whole discussion shows that I like to discuss.

Which is excellent and much appreciated.

>  I am sure I have a long road ahead toward perfection (who doesn’t except our Jedi?).

Your Jedi has a long way to go too.

**Comment by jacg** [comment]
My `test.matplotlib` [branch](https://github.com/jacg/IC/tree/test.matplotlib) shows a brief outline. @jjgomezcadenas Please try it on your Mac, to see how the image compares there. On Linux, the python 3.6 version seems to put the y-axis labels a bit higher than in the older versions. I'd be interested to see if the same happens on OS X.

Also, the Agg-selection doesn't seem to be working, so it's crashing on headless environments (including Travis), even though it appeared that I had that problem fixed locally for a while (it seems to be back).

Still, it's a start.

**Comment by jjgomezcadenas** [comment]
Tried in the MAC, works fine. I agree, this gives us a way to add tests to plot functions. 

Could you send a PR for this example? I think it is very useful and must be in the repository.

**Comment by jacg** [comment]
Did you check whether it behaves consistently across all our supported python versions? (Our exchange on another issue leads me to suspect that you have not.)

While this is a nice idea, on my machine 3.6 gives slightly different picutres from 3.5 and 2.7. As such we get spurious failures. This will have to be fixed before being accepted.

Secondly, I seem to recall that there were problems with it on Travis. These will also need fixing. I'll make a PR if you promise *not* to accept it until these issues are resolved! It's perfectly OK open a PR before it is ready to be accepted, and leave it open while it is being worked on. Indeed, that is what you should have done with #111 rather than closing it and opening #112 instead. Now we have 2 PR addressing the same issue, and the discussion relating to it is spread between two places.

**Comment by jacg** [comment]
The problems in my `test.matplotlib` [branch](https://github.com/jacg/IC/tree/test.matplotlib) seem to have gone away. By this I mean

+ Everything passes in Travis
+ Everything passes locally, as long as I recreate the `IC3.5` and `IC3.6` conda environments from scratch.

In other words, the current versions of the relevant libraries no longer seem to suffer from the two problems we observed:

+ Slight difference in y-axis label heights between Python 3.6 and others.
+ Inability to switch to Agg backend before starting tests.

If this really does work, and I'm not merely overlooking something, then I think that we should add a test like this to `master`. I would prefer that be a real test, rather than the toy example used in this branch.

Please suggest candidates for this test.

The reference images should really go in LFS rather than being versioned directly in Git, which increases the importance of #210.

**Comment by jacg** [comment]
Yet another item stalled, waiting for #210 (Git LFS).

**Comment by jacg** [comment]
I'm closing this in order to clean up the Kanban board, as it has been implemented in #119.
